drivers/crypto/crypto_esp32_sha.c

```diff
@@ -0,0 +1,583 @@
+		return;
+	}
+
+	s->hal_mode = params.hal_mode;
+	s->block_bytes = params.block_bytes;
+	s->state_words = params.state_words;
+	s->out_bytes = params.out_bytes;
+	s->len_field_bytes = params.len_field_bytes;
+
+	if (algo == CRYPTO_HASH_ALGO_SHA224) {
+		memcpy(s->H, sha224_init_state, sizeof(sha224_init_state));
+	}
+}
+
+static struct esp_sha_ctx *sha_pool_alloc(enum hash_algo algo)
+{
+	struct esp_sha_ctx *ret = NULL;
+
+	k_mutex_lock(&sha_pool_lock, K_FOREVER);
+
+	for (size_t i = 0; i < ARRAY_SIZE(sha_pool); i++) {
+		if (!sha_pool[i].in_use) {
+			sha_pool[i].in_use = true;
+			sha_ctx_init_params(&sha_pool[i], algo);
+			ret = &sha_pool[i];
+			break;
+		}
+	}
+
+	k_mutex_unlock(&sha_pool_lock);
+
+	if (!ret) {
+		LOG_WRN("No available SHA context in pool");
+	}
+
+	return ret;
+}
+
+static void sha_pool_free(struct esp_sha_ctx *s)
+{
+	if (!s) {
+		return;
+	}
+
+	k_mutex_lock(&sha_pool_lock, K_FOREVER);
+	memset(s, 0, sizeof(*s));
+	k_mutex_unlock(&sha_pool_lock);
+}
+
+#if SOC_SHA_SUPPORT_RESUME
+static inline void sha_hw_restore(struct esp_sha_ctx *s)
+{
+	if (s->first_block && s->algo != CRYPTO_HASH_ALGO_SHA224) {
+		return;
+	}
+
+	sha_hal_write_digest(s->hal_mode, s->H);
+}
+
+#else  /* !SOC_SHA_SUPPORT_RESUME */
+
+static inline void sha_ll_write_digest_esp32(esp_sha_type sha_type, void *digest_state,
+					     size_t state_len)
+{
+	uint32_t *digest_state_words = (uint32_t *)digest_state;
+	uint32_t *reg_addr_buf = (uint32_t *)(SHA_TEXT_BASE);
+
+	if (sha_type == SHA2_384 || sha_type == SHA2_512) {
+		for (size_t i = 0; i < state_len; i += 2) {
+			reg_addr_buf[i] = digest_state_words[i + 1];
+			reg_addr_buf[i + 1] = digest_state_words[i];
+		}
+	} else {
+		for (size_t i = 0; i < state_len; i++) {
+			reg_addr_buf[i] = digest_state_words[i];
+		}
+	}
+
+	sha_ll_load(sha_type);
+}
+
+static inline void sha_hw_restore_esp32(struct esp_sha_ctx *s)
+{
+	if (s->first_block && s->algo != CRYPTO_HASH_ALGO_SHA224) {
+		return;
+	}
+
+	sha_ll_write_digest_esp32(s->hal_mode, s->H, s->state_words);
+}
+#endif /* SOC_SHA_SUPPORT_RESUME */
+
+static size_t sha_make_padding(const struct esp_sha_ctx *s, const uint8_t *tail, size_t tail_len,
+			       uint8_t last[128], uint8_t last2[128])
+{
+	const size_t B = s->block_bytes;
+	const size_t L = s->len_field_bytes;
+	const uint64_t bit_len = s->total_len * 8ULL;
+
+	if (tail_len >= B) {
+		LOG_ERR("Invalid tail length: %zu", tail_len);
+		return 0;
+	}
+
+	memset(last, 0, B);
+	if (tail_len > 0) {
+		memcpy(last, tail, tail_len);
+	}
+	last[tail_len] = 0x80;
+
+	if (tail_len + 1 + L <= B) {
+		if (L == 16) {
+			memset(&last[B - 16], 0, 8);
+		}
+
+		for (int i = 0; i < 8; i++) {
+			last[B - 8 + i] = (uint8_t)(bit_len >> (56 - 8 * i));
+		}
+		return 1;
+	}
+
+	memset(last2, 0, B);
+	if (L == 16) {
+		memset(&last2[B - 16], 0, 8);
+	}
+	for (int i = 0; i < 8; i++) {
+		last2[B - 8 + i] = (uint8_t)(bit_len >> (56 - 8 * i));
+	}
+
+	return 2;
```

Comment:
single comment about default return value `2` would help here

Reply:
It is just the number of paddings, nothing else.